**2. Given an unsorted array of integers, find the length of the longest continuous increasing subsequence (subarray).**


Example 1:
Input: [1,3,5,4,7]
Output: 3 


Example 2:
Input: [2,2,2,2,2]
Output: 1


In [6]:
def LLCIS(nums):
    if not nums:
        return 0
    
    longest_length = 1
    current_length = 1
    
    for i in range(1, len(nums)):
        if nums[i] > nums[i - 1]:
            current_length += 1
        else:
            if current_length > longest_length:
                longest_length = current_length
            current_length = 1
    
    if current_length > longest_length:
        longest_length = current_length
    
    return longest_length

# Example usage:
Sample_input1=[1, 3, 5, 4, 7]
Sample_input2=[2, 2, 2, 2, 2]

#User Input code
#S=list(map(int,input().split(",")))


print(LLCIS(Sample_input1))
print(LLCIS(Sample_input2))


3
1


---------------------------------------------------------------------------------------------------------------------------------------------------------

**3. Given a list of non negative integers, arrange them such that they form the largest number.**
 
Example 1:
Input: [10,2]
Output: "210"


Example 2:
Input: [3,30,34,5,9]
Output: "9534330"


In [11]:
def largestNumber(array):
    #If there is only one element in the list, the element itself is the largest element. 
    if len(array)==1:  
        return str(array[0])
    
    #convert a list into a string array.
    for i in range(len(array)):
        array[i]=str(array[i])


    #find the largest element by swapping technique.
    for i in range(len(array)):
        for j in range(1+i,len(array)):
            if array[j]+array[i]>array[i]+array[j]:
                array[i],array[j]=array[j],array[i]


    #JOIN function in Python
    result=''.join(array)


    #If all elements are 0, answer must be 0 
    if(result=='0'*len(result)):
        return '0'
    else:
        return result
        
# Example usage:
Sample_input1=[10,2]
Sample_input2=[3,30,34,5,9]


#User Input code
#S=list(map(int,input().split(",")))


print(largestNumber(Sample_input1))
print(largestNumber(Sample_input2))


210
9534330


--------------------------------------------------------------------------------------------------------------------------------------------------------

**4. Store all the "servlet-name", and "servlet-class" to a csv file from the attached sample_json.json file using Python.**

In [28]:
import json
import csv

Path="C:\\Users\\Hp\\Desktop\\adept ready\\DT A1 sample_json.json"
# Load JSON data from file
with open(Path) as f:
    data = json.load(f)

# Extract "servlet-name" and "servlet-class" from the JSON data
servlets = []
for servlet in data['web-app']['servlet']:
    servlet_name = servlet['servlet-name']
    servlet_class = servlet['servlet-class']
    servlets.append((servlet_name, servlet_class))

# Write to CSV file
with open('servlet_data.csv', 'w', newline='') as csvfile:
    fieldnames = ['servlet-name', 'servlet-class']
    
    # Create a CSV DictWriter object, specifying the field names
    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
    
    # Write the header row containing the field names
    writer.writeheader()

    # Iterate over each servlet extracted from the JSON data
    for servlet in servlets:
        writer.writerow({'servlet-name': servlet[0], 'servlet-class': servlet[1]})

---------------------------------------------------------------------------------------------------------------------

**1.Data set reference link: https://www.consumerfinance.gov/data-research/consumer-complaints/#download-the-data
File data source: https://files.consumerfinance.gov/ccdb/complaints.csv.zip 
Problem statement:
Download the data from the file data source and provide possible data insights.**


**DATA PREPROCESSING**

In [18]:
import pandas as pd

Path="C:\\Users\\Hp\\Desktop\\adept ready\\complaints-2024-05-09_02_30.csv"

# Read CSV file into a Pandas DataFrame
df = pd.read_csv(Path)


In [28]:
# checking Data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1552333 entries, 0 to 1552332
Data columns (total 18 columns):
 #   Column                        Non-Null Count    Dtype  
---  ------                        --------------    -----  
 0   Date received                 1552333 non-null  object 
 1   Product                       1552333 non-null  object 
 2   Sub-product                   1552331 non-null  object 
 3   Issue                         1552331 non-null  object 
 4   Sub-issue                     1518925 non-null  object 
 5   Consumer complaint narrative  452976 non-null   object 
 6   Company public response       773839 non-null   object 
 7   Company                       1552333 non-null  object 
 8   State                         1548005 non-null  object 
 9   ZIP code                      1552281 non-null  object 
 10  Tags                          87945 non-null    object 
 11  Consumer consent provided?    1265770 non-null  object 
 12  Submitted via               

In [19]:
# Shape of the DataFrame
df.shape

(1552333, 18)

In [20]:
# Display the DataFrame
df.head()

Date received                                            Product  \
0      03/29/24  Credit reporting or other personal consumer re...   
1      04/06/24  Credit reporting or other personal consumer re...   
2      03/23/24  Credit reporting or other personal consumer re...   
3      03/07/24  Credit reporting or other personal consumer re...   
4      03/25/24  Credit reporting or other personal consumer re...   

        Sub-product                                              Issue  \
0  Credit reporting  Problem with a company's investigation into an...   
1  Credit reporting  Problem with a company's investigation into an...   
2  Credit reporting                        Improper use of your report   
3  Credit reporting  Problem with a company's investigation into an...   
4  Credit reporting  Problem with a company's investigation into an...   

                                           Sub-issue  \
0  Their investigation did not fix an error on yo...   
1  Their investigation did not fix an error on yo...   
2      Reporting company used your report improperly   
3  Their investigation did not fix an error on yo...   
4  Was not notified of investigation status or re...   

                        Consumer complaint narrative  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  I disputed a list of accounts that were not ve...   
4  I have consistently maintained on-time payment...   

                             Company public response  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  Company has responded to the consumer and the ...   
4  Company has responded to the consumer and the ...   

                                  Company State ZIP code Tags  \
0                           EQUIFAX, INC.    NY    10034  NaN   
1     Experian Information Solutions Inc.    AR    72204  NaN   
2             Early Warning Services, LLC    MS    39759  NaN   
3  TRANSUNION INTERMEDIATE HOLDINGS, INC.    NY    XXXXX  NaN   
4  TRANSUNION INTERMEDIATE HOLDINGS, INC.    IL    60440  NaN   

  Consumer consent provided? Submitted via Date sent to company  \
0                        NaN           Web             03/29/24   
1                        NaN           Web             04/06/24   
2                        NaN           Web             03/23/24   
3           Consent provided           Web             03/07/24   
4           Consent provided           Web             03/25/24   

      Company response to consumer Timely response?  Consumer disputed?  \
0                      In progress              Yes                 NaN   
1                      In progress              Yes                 NaN   
2          Closed with explanation              Yes                 NaN   
3  Closed with non-monetary relief              Yes                 NaN   
4  Closed with non-monetary relief              Yes                 NaN   

   Complaint ID  
0       8653977  
1       8710328  
2       8611411  
3       8501851  
4       8617108

**We have extracted data from the past year, resulting in a dataset comprising 1,552,333 rows and 18 columns.**

In [25]:
#Duplicate value identification
duplicate_rows = df[df.duplicated()]
print(len(duplicate_rows))

0


**There is no duplicate rows presented in dataset.**

In [26]:
#Null value identification
df.isna().sum()

Date received                         0
Product                               0
Sub-product                           2
Issue                                 2
Sub-issue                         33408
Consumer complaint narrative    1099357
Company public response          778494
Company                               0
State                              4328
ZIP code                             52
Tags                            1464388
Consumer consent provided?       286563
Submitted via                         0
Date sent to company                  0
Company response to consumer         12
Timely response?                      0
Consumer disputed?              1552333
Complaint ID                          0
dtype: int64


Consumer complaint narrative: 1,099,357 missing values.


Company public response: 778,494 missing values.


Tags: 1,464,388 missing values.


Consumer consent provided?: 286,563 missing values.


Consumer disputed?: 1,552,333 missing values.



**These are the few columns has highest count of Nan values.Since we dropping those columns for getting better insights about the data** 
**NOTE: Kindly visit the link attached below to know about each columns**


https://www.consumerfinance.gov/complaint/data-use/

In [34]:
df.drop(columns=["Consumer complaint narrative","Tags","Consumer consent provided?",
                 "Company public response","Consumer disputed?"],inplace=True)

In [39]:
# Replace null values in specific columns with 'unknown'
columns_to_replace = ["Issue","Sub-product","Sub-issue","State","ZIP code","Company response to consumer"]
df[columns_to_replace] = df[columns_to_replace].fillna('unknown')


In [42]:
df.shape

(1552333, 13)

**After preprocessing we got dataset of 1,552,333 rows and 13 columns.**

In [44]:
df.isna().sum()

Date received                   0
Product                         0
Sub-product                     0
Issue                           0
Sub-issue                       0
Company                         0
State                           0
ZIP code                        0
Submitted via                   0
Date sent to company            0
Company response to consumer    0
Timely response?                0
Complaint ID                    0
dtype: int64

In [92]:

file_path = 'complaints_data.csv'
df.to_csv(file_path, index=False)

**EXPORTED AS complaints_data.csv FILE**